File: parser.js
import { TokenType } from "./lexer.js";

class NumberNode {
	constructor(value) {
		this.value = value;
	}

	evaluate() {
		return this.value;
	}

	toString() {
		return `NumberNode(${this.value})`;
	}
}

class BinaryOpNode {
	constructor(left, op, right) {
		this.left = left;
		this.op = op;
		this.right = right;
	}

	evaluate() {
		switch (this.op.type) {
			case TokenType.PLUS:
				return this.left.evaluate() + this.right.evaluate();
			case TokenType.MINUS:
				return this.left.evaluate() - this.right.evaluate();
			case TokenType.TIMES:
				return this.left.evaluate() * this.right.evaluate();
			case TokenType.DIVIDE:
				return this.left.evaluate() / this.right.evaluate();
			default:
				throw new Error("Invalid operation");
		}
	}

	toString() {
		return `BinaryOpNode(${this.left.toString()}, ${this.op.type}, ${this.right.toString()})`;
	}
}

class Parser {
	constructor(tokens) {
		this.tokens = tokens;
		this.pos = 0;
		this.currentToken = this.tokens[this.pos];
	}

	error() {
		throw new Error("Invalid syntax");
	}

	eat(tokenType) {
		if (this.currentToken.type === tokenType) {
			this.pos += 1;
			if (this.pos < this.tokens.length) {
				this.currentToken = this.tokens[this.pos];
			}
		} else {
			this.error();
		}
	}

	parse() {
		return this.expr();
	}

	expr() {
		let node = this.term();
		while ([TokenType.PLUS, TokenType.MINUS].includes(this.currentToken.type)) {
			const token = this.currentToken;
			if (token.type === TokenType.PLUS) {
				this.eat(TokenType.PLUS);
			} else if (token.type === TokenType.MINUS) {
				this.eat(TokenType.MINUS);
			}
			node = new BinaryOpNode(node, token, this.term());
		}
		return node;
	}

	term() {
		let node = this.factor();
		while (
			[TokenType.TIMES, TokenType.DIVIDE].includes(this.currentToken.type)
		) {
			const token = this.currentToken;
			if (token.type === TokenType.TIMES) {
				this.eat(TokenType.TIMES);
			} else if (token.type === TokenType.DIVIDE) {
				this.eat(TokenType.DIVIDE);
			}
			node = new BinaryOpNode(node, token, this.factor());
		}
		return node;
	}

	factor() {
		const token = this.currentToken;
		if (token.type === TokenType.NUMBER) {
			this.eat(TokenType.NUMBER);
			return new NumberNode(token.value);
		}
		if (token.type === TokenType.LPAREN) {
			this.eat(TokenType.LPAREN);
			const node = this.expr();
			this.eat(TokenType.RPAREN);
			return node;
		}
		this.error();
	}
}

import { ASTToBytecodeConverter } from "./ast-to-bytecode-converter.js";

export { Parser, NumberNode, BinaryOpNode };

File: parser.py
from lexer import TokenType

class NumberNode:
    def __init__(self, value):
        self.value = value

class BinaryOpNode:
    def __init__(self, left, op, right):
        self.left = left
        self.op = op
        self.right = right

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0
        self.current_token = self.tokens[self.pos]

    def error(self):
        raise Exception("Invalid syntax")

    def eat(self, token_type):
        if self.current_token.type == token_type:
            self.pos += 1
            if self.pos < len(self.tokens):
                self.current_token = self.tokens[self.pos]
        else:
            self.error()

    def parse(self):
        return self.expr()

    def expr(self):
        node = self.term()
        while self.current_token.type in (TokenType.PLUS, TokenType.MINUS):
            token = self.current_token
            if token.type == TokenType.PLUS:
                self.eat(TokenType.PLUS)
            elif token.type == TokenType.MINUS:
                self.eat(TokenType.MINUS)
            node = BinaryOpNode(left=node, op=token, right=self.term())
        return node

    def term(self):
        node = self.factor()
        while self.current_token.type in (TokenType.TIMES, TokenType.DIVIDE):
            token = self.current_token
            if token.type == TokenType.TIMES:
                self.eat(TokenType.TIMES)
            elif token.type == TokenType.DIVIDE:
                self.eat(TokenType.DIVIDE)
            node = BinaryOpNode(left=node, op=token, right=self.factor())
        return node

    def factor(self):
        token = self.current_token
        if token.type == TokenType.NUMBER:
            self.eat(TokenType.NUMBER)
            return NumberNode(value=token.value)
        elif token.type == TokenType.LPAREN:
            self.eat(TokenType.LPAREN)
            node = self.expr()
            self.eat(TokenType.RPAREN)
            return node
        else:
            self.error()
